tsfc/coffee_mode.py

```diff
@@ -95,6 +93,44 @@ def monomial_sum_to_expression(monomial_sum):
     return make_sum(indexsums)
 
 
+def solve_ip(variables, is_feasible, compare):
+    """Solve a 0-1 integer programming problem. The algorithm tries to set each
+    variable to 1 recursively, and stop the recursion early by comparing with
+    the optimal solution at entry. At worst case this is 2^N (as this is a
+    NP-hard problem), but recursions can be trimmed as soon as a reasonable
+    solution have been found.
+
+    :param varialbes: list of unique 0-1 variables
+    :param is_feasible: function to test if a combination is feasible
+    :param compare: function to compare two solutions, compare(s1, s2) returns
+           True if s1 is a better solution than s2, and False otherwise
```

Comment:
You call it compare, but returning a logical value, it is a *strictly less* operator. I still think you should go the Python way, and have a key function instead.